File: courseMaterial/course_ROSActions/code_section_3_simple_actions_activity/my_robot_tutorials/src/move_robot_server.cpp
#include <ros/ros.h>
#include <actionlib/server/simple_action_server.h>

#include <string>

#include <my_robot_msgs/MoveRobotAction.h>
#include <my_robot_msgs/MoveRobotGoal.h>
#include <my_robot_msgs/MoveRobotResult.h>
#include <my_robot_msgs/MoveRobotFeedback.h>

class MoveRobotServer {

	protected:
	ros::NodeHandle _nh;
	actionlib::SimpleActionServer<my_robot_msgs::MoveRobotAction> _as;
	int _current_position;

	public:
	MoveRobotServer():
		_as(_nh, "/move_robot",
		boost::bind(&MoveRobotServer::onGoal, this, _1),
		false),
		_current_position(50)
	{
		_as.start();
		ROS_INFO("Move Robot server has been started");
	}

	void sendFeedback()
	{
		my_robot_msgs::MoveRobotFeedback feedback;
		feedback.current_position = _current_position;
		_as.publishFeedback(feedback);
	}

	void onGoal(const my_robot_msgs::MoveRobotGoalConstPtr &goal)
	{
		ROS_INFO("Goal received");
		int goal_position = goal->position;
		int velocity = goal-> velocity;
		ROS_INFO("Goal position: %d, Velocity: %d", 
			goal_position, velocity);

		bool success = false;
		bool preempted = false;
		bool invalid_parameters = false;
		std::string message = "";
		ros::Rate rate(1.0);

		if (goal_position < 0 || goal_position > 100) {
			message = "Invalid goal position, must be [0-100]";
			invalid_parameters = true;
		}

		if (goal_position == _current_position) {
			success = true;
			message = "Current position is already correct!";
		}

		while (ros::ok() && !success && !invalid_parameters) {
			if (_as.isPreemptRequested()) {
				if (goal_position == _current_position) {
					message = "Preempted but already at goal position";
					success = true;
				}
				else {
					message = "Preempted and stopped execution";
					preempted = true;
				}
				break;
			}

			int diff = goal_position - _current_position;

			if (diff == 0) {
				message = "Success";
				success = true;
				break;
			}
			else if (diff < 0) {
				if (abs(diff) >= velocity) {
					_current_position -= velocity;
				}
				else {
					_current_position -= abs(diff);
				}
			}
			else if (diff > 0) {
				if (diff >= velocity) {
					_current_position += velocity;
				}
				else {
					_current_position += diff;
				}
			}

			sendFeedback();
			rate.sleep();
		}

		my_robot_msgs::MoveRobotResult result;
		result.position = _current_position;
		result.message = message;
		ROS_INFO("Send goal result to client");

		if (preempted) {
			ROS_INFO("Preempted");
			_as.setPreempted(result);
		}
		else if (success) {
			ROS_INFO("Success");
			_as.setSucceeded(result);
		}
		else {
			ROS_INFO("Aborted");
			_as.setAborted(result);
		}
	}
};

int main (int argc, char **argv) {
	ros::init(argc, argv, "move_robot_server");

	MoveRobotServer server;
	ros::spin();
}

File: courseMaterial/course_ROSActions/code_section_5_actions_activity/my_robot_tutorials/scripts/move_robot_server.py
#!/usr/bin/env python

import rospy
import actionlib
from my_robot_msgs.msg import MoveRobotAction
from my_robot_msgs.msg import MoveRobotFeedback
from my_robot_msgs.msg import MoveRobotResult

class MoveRobotServer:

    def __init__(self):
        self._as = actionlib.SimpleActionServer('/move_robot',
            MoveRobotAction, execute_cb=self.on_goal, auto_start=False)
        self._as.start()

        self._current_position = 50
        rospy.loginfo("Server has been started")
    
    def send_feedback(self):
        feedback = MoveRobotFeedback()
        feedback.current_position = self._current_position
        self._as.publish_feedback(feedback)

    def on_goal(self, goal):
        rospy.loginfo("A goal has been received!")
        rospy.loginfo(goal)

        goal_position = goal.position
        velocity = goal.velocity

        success = False
        preempted = False
        invalid_parameters = False
        message = ""
        rate = rospy.Rate(1.0)

        if goal_position < 0 or goal_position > 100:
            message = "Invalid goal position, must be [0-100]"
            invalid_parameters = True

        if goal_position == self._current_position:
            success = True
            message = "Current position is already correct!"

        while not rospy.is_shutdown() and not success \
            and not invalid_parameters:
            if self._as.is_preempt_requested():
                if goal_position == self._current_position:
                    message = "Preempted but already at goal position"
                    success = True
                else:
                    message = "Preempted and stopped execution"
                    preempted = True
                break

            diff = goal_position - self._current_position

            if diff == 0:
                message = "Success"
                success = True
                break
            elif diff < 0:
                if abs(diff) >= velocity:
                    self._current_position -= velocity
                else:
                    self._current_position -= abs(diff)
            elif diff > 0:
                if diff >= velocity:
                    self._current_position += velocity
                else:
                    self._current_position += diff
            
            self.send_feedback()
            rate.sleep()

        result = MoveRobotResult()
        result.position = self._current_position
        result.message = message
        rospy.loginfo("Send goal result to client")

        if preempted:
            rospy.loginfo("Preemted")
            self._as.set_preempted(result)
        elif success:
            rospy.loginfo("Success")
            self._as.set_succeeded(result)
        else:
            rospy.loginfo("Aborted")
            self._as.set_aborted(result)


if __name__ == '__main__':
    rospy.init_node('move_robot_server')

    server = MoveRobotServer()

    rospy.spin()

